ask-sdk-s3-persistence-adapter/lib/attributes/persistence/S3PersistenceAdapter.ts

```diff
@@ -0,0 +1,106 @@
+ * permissions and limitations under the License.
+ */
+
+import { PersistenceAdapter } from 'ask-sdk-core';
+import { RequestEnvelope } from 'ask-sdk-model';
+import { S3 } from 'aws-sdk';
+import { createAskSdkError } from '../../utils/AskSdkUtils';
+import {
+    ObjectKeyGenerator,
+    ObjectKeyGenerators,
+} from './ObjectKeyGenerators';
+
+/**
+ * Implementation of {@link PersistenceAdapter} using AWS S3
+ */
+export class S3PersistenceAdapter implements PersistenceAdapter {
+    protected bucketName : string;
+    protected s3Client : S3;
+    protected objectKeyGenerator : ObjectKeyGenerator;
+
+    constructor(config : {
+        bucketName : string,
+        s3Client? : S3,
+        objectKeyGenerator? : ObjectKeyGenerator,
+    }) {
+        this.bucketName = config.bucketName;
+        this.s3Client = config.s3Client ? config.s3Client : new S3({apiVersion : 'latest'});
+        this.objectKeyGenerator = config.objectKeyGenerator ? config.objectKeyGenerator : ObjectKeyGenerators.userId;
+    }
+
+    /**
+     * Retrieves persistence attributes from AWS S3.
+     * @param {RequestEnvelope} requestEnvelope Request envelope used to generate object key.
+     * @returns {Promise<Object.<string, any>>}
+     */
+    public async getAttributes(requestEnvelope : RequestEnvelope) : Promise<{[key : string] : string}> {
+        const objectId = this.objectKeyGenerator(requestEnvelope);
+
+        const getParams : S3.GetObjectRequest = {
+            Bucket : this.bucketName,
+            Key : objectId,
+        };
+
+        let data : S3.GetObjectOutput;
+
+        try {
+            data = await this.s3Client.getObject(getParams).promise();
+        } catch (err) {
+            if (err.code === 'NoSuchKey') {
+                return {};
+            }
+
+            throw createAskSdkError(
+                this.constructor.name,
+                `Could not read item (${objectId}) from bucket (${getParams.Bucket}): ${err.message}`,
+            );
+        }
+
+        const bodyString = data.Body ? data.Body.toString() : '';
+
+        let bodyObj;
+
+        try {
+            bodyObj = bodyString ? JSON.parse(bodyString) : {};
+        } catch (err) {
+            throw new SyntaxError(`Failed trying to parse the data body: ${data.Body.toString()}`);
+        }
+
+        return bodyObj;
+    }
+
+    /**
+     * Saves persistence attributes to AWS S3.
+     * @param {RequestEnvelope} requestEnvelope Request envelope used to generate object key.
+     * @param {Object.<string, any>} attributes Attributes to be saved to DynamoDB.
+     * @return {Promise<void>}
+     */
+    public async saveAttributes(requestEnvelope : RequestEnvelope, attributes : {[key : string] : string}) : Promise<void> {
+        const objectId = this.objectKeyGenerator(requestEnvelope);
+
+        const putParams : S3.PutObjectRequest = {
+            Bucket : this.bucketName,
+            Key : objectId,
+            Body : JSON.stringify(attributes),
```

Comment:
Can ``JSON.stringify`` throw an error that needs to be caught?

Reply:
Nope :D it will just return the original string if it's not json